0005-kinesis-proxy-json.md

```diff
@@ -0,0 +1,306 @@
+
+## Further Documentation
+
+- Kinesis high-level overview: https://docs.aws.amazon.com/streams/latest/dev/key-concepts.html
+- Kinesis API reference: https://docs.aws.amazon.com/kinesis/latest/APIReference/Welcome.html
+- Kinesis Enhanced Fan-out Consumer: https://docs.aws.amazon.com/streams/latest/dev/building-enhanced-consumers-api.html
+- `ex_aws_kinesis` (ExAws interface to the Kinesis API): https://docs.aws.amazon.com/kinesis/latest/APIReference/Welcome.html
+- Kinesis Client Library (KCL): https://docs.aws.amazon.com/streams/latest/dev/shared-throughput-kcl-consumers.html
+- Elixir Multi-Lang Daemon interface: https://github.com/AdRoll/exmld
+- `kcl_ex` (native Elixir KCL implementation): https://github.com/uberbrodt/kcl_ex
+- JSON Schema: https://jsonschema.org
+
+# Drawbacks
+
+Kinesis is a new technology that we're adding to our infrastructure. Even
+though it's mostly managed by Amazon, we'll need to develop some in-house
+expertise with building producers and consumers.
+
+Enhanced fan-out consumers are not well-supported by the existing
+Elixir/Erlang Kinesis adapters. The functionality does work (Paul Swartz was
+able to get it working during proof-of-concept testing) but we'll need to
+write some of our own code to support the functionality.
+
+Running Kinesis locally is easy enough with tools like LocalStack or
+Kinesialite, but more involved than running a TCP proxy. This might require
+some re-working of how the RTR integration system works when pushing logged
+data into a local instance for testing. Kinesialite doesn't support enhanced
+fan-out, so we'd need to support both models in our code.
+
+# Rationale and alternatives
+
+<!-- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this? -->
+
+AWS has tools to proxy messages from Kinesis into Kafka (and vice versa) so if
+we need to migrate in the future that is possible. If we had a lot more
+operational staff (or more experience with Kafka) this might point to a
+different decision, but I believe that Kinesis gets us the value we need with
+the minimum of cost and operational overhead.
+
+## Managed Streaming for Kafka (MSK)
+
+MSK is the main managed alternative to Kinesis in
+AWS. [Kafka](https://kafka.apache.org/) is "an open-source distributed event
+streaming platform".
+
+### Advantages
+- more external software written to write to/consume Kafka
+- vendor-neutral: Kafka can run on any cloud or on-premises
+- pricing is based on cluster size and storage, so not as closely tied to the
+  number of events
+- messages have additional metadata, requiring slightly less overhead when
+  using CloudEvents
+- Being used by Cubic for Fare Transformation in Azure
+
+### Disadvantages
+- MSK cannot currently resize a cluster, so we either need to pay more up
+  front, or pay an additional migration cost as we scale up
```

Comment:
Is "scaling up" much of a concern within the scope of this RFC? I'd expect OCS message volume to be roughly constant over time.